source/extensions/filters/listener/http_inspector/http_inspector.cc

```diff
@@ -0,0 +1,145 @@
+#include "extensions/filters/listener/http_inspector/http_inspector.h"
+
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/listen_socket.h"
+#include "envoy/stats/scope.h"
+
+#include "common/api/os_sys_calls_impl.h"
+#include "common/common/assert.h"
+#include "common/common/macros.h"
+#include "common/http/headers.h"
+
+#include "extensions/transport_sockets/well_known_names.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace ListenerFilters {
+namespace HttpInspector {
+
+Config::Config(Stats::Scope& scope)
+    : stats_{ALL_HTTP_INSPECTOR_STATS(POOL_COUNTER_PREFIX(scope, "http_inspector."))} {}
+
+const absl::string_view Filter::HTTP2_CONNECTION_PREFACE = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+thread_local uint8_t Filter::buf_[Config::MAX_INSPECT_SIZE];
+
+Filter::Filter(const ConfigSharedPtr config) : config_(config) {}
+
+Network::FilterStatus Filter::onAccept(Network::ListenerFilterCallbacks& cb) {
+  ENVOY_LOG(debug, "http inspector: new connection accepted");
+
+  Network::ConnectionSocket& socket = cb.socket();
+
+  absl::string_view transport_protocol = socket.detectedTransportProtocol();
+  if (!transport_protocol.empty() &&
+      transport_protocol != TransportSockets::TransportSocketNames::get().RawBuffer) {
+    ENVOY_LOG(trace, "http inspector: cannot inspect http protocol with transport socket {}",
+              transport_protocol);
+    return Network::FilterStatus::Continue;
+  }
+
+  ASSERT(file_event_ == nullptr);
+
+  file_event_ = cb.dispatcher().createFileEvent(
+      socket.ioHandle().fd(),
+      [this](uint32_t events) {
+        ASSERT(events == Event::FileReadyType::Read);
+        onRead();
+      },
+      Event::FileTriggerType::Edge, Event::FileReadyType::Read);
+
+  cb_ = &cb;
+  return Network::FilterStatus::StopIteration;
+}
+
+void Filter::onRead() {
+  auto& os_syscalls = Api::OsSysCallsSingleton::get();
+  const Api::SysCallSizeResult result =
+      os_syscalls.recv(cb_->socket().ioHandle().fd(), buf_, Config::MAX_INSPECT_SIZE, MSG_PEEK);
+  ENVOY_LOG(trace, "http inspector: recv: {}", result.rc_);
+  if (result.rc_ == -1 && result.errno_ == EAGAIN) {
+    return;
+  } else if (result.rc_ < 0) {
+    config_->stats().read_error_.inc();
+    done(false);
+    return;
+  }
+
+  parseHttpHeader(absl::string_view(reinterpret_cast<const char*>(buf_), result.rc_));
+}
+
+void Filter::parseHttpHeader(absl::string_view data) {
+  if (absl::StartsWith(data, Filter::HTTP2_CONNECTION_PREFACE)) {
+    ENVOY_LOG(trace, "http inspector: http2 connection preface found");
+    protocol_ = "HTTP/2";
+    done(true);
+  } else {
+    size_t pos = data.find_first_of("\r\n");
+
+    // Cannot find \r or \n
+    if (pos == absl::string_view::npos) {
+      ENVOY_LOG(trace, "http inspector: no request line detected");
+      return done(false);
```

Comment:
Erm, this won't work. This whole function assumes that everything will arrive in first read, which is definitely not always the case (e.g. you may receive input byte-by-byte), and which will lead to false negatives.

Reply:
This may cause the inspector to wait infinitely (e.g., client sends `GET /index` only). Do you think we should add a timer to handle this case or is there any other way to avoid the wait?

I found `proxy protocol` filter might have the same problem.

Reply:
There is already a global timeout for all listener filters. See #5231.

Reply:
Ugh, I misread this and thought this will wait for another read...
@PiotrSikora is right `done(false)` is not appropriate here. You should treat this like the handling of SSL_WANT_READ in tls_inspector.

Reply:
In addition to that, if the first read is partial HTTP2 connection preface (e.g. `"PRI * HTTP/2.0\r\n"`), this function falls into HTTP/1 path and will call `done(false)` too.

Reply:
That global timer simply terminates the connection. It doesn't pass the connection to the default TCP filter stack. Its the same issue with TLS inspector as well

Reply:
@rshriram that's a separate issue, which is independent of this inspector. #7195